### ISSUE TITLE
CUDA support

### ISSUE THREAD

**Issue by bentsherman** [SystemsGenetics/ACE#25]
This issue is very low-priority at the moment, but in case I ever need to explore potential performance benefits with CUDA, I'd like to have thought out what I will and won't be able to do with CUDA and ACE. That is, whether it is possible to implement a CUDA wrapper in ACE similar to the OpenCL wrapper. I really like the system of blocks for OpenCL and so I'd like to be able to replicate this structure for CUDA rather than call CUDA kernels in `runSerial()`, if it is possible.

The primary difference is that CUDA kernels are compiled with the source code, whereas OpenCL kernels are simply included as plain-text "resources" in the binary and compiled at run-time. We can't compile CUDA kernels with a Qt application, so instead __all .cu files must be compiled separately with `nvcc` and included as object files in the Qt project file__. Furthermore, __each kernel must be accompanied by a host wrapper function__, because the syntax for executing a kernel can only be compiled by `nvcc`.

Given the compiled CUDA kernels, the rest of the CUDA API consists of host functions, so they can be mixed in with Qt. You select a CUDA device (although the first one is selected automatically), copy any data to the device, execute the kernel (or in our case a pre-compiled function which calls the kernel), and then copy any data from the device. It's actually much simpler than equivalent OpenCL code, and I think it can be done entirely through a C/C++ API like OpenCL.

Maybe once MPI support is done, and once OpenCL and MPI are optional features, I will see if I can implement this myself.

**Comment by bentsherman** [comment]
Well, with no other pressing tasks at the moment I went ahead and tried this out, and so far so good! I have a "Select CUDA Device" dialog alongside the OpenCL equivalent, and I can view and select devices.

**Comment by bentsherman** [comment]
Saved my progress on this feature to the `cuda-support` branch. I think all of the necessary infrastructure is there now, so I will add OpenCL and CUDA code to the example program to demonstrate their usage.

As it stands, I just extend `runBlock()` so that it can run OpenCL or CUDA kernels, and ACE manages either one with the same code. So the `runBlock()` implementation of a real analytic might look like this:
```
   if ( capability == Capabilities::OpenCL ) {
      // run OpenCL...
   }
   else if ( capability == Capabilities::CUDA ) {
      // run CUDA...
   }
```

Now let's just hope I can pre-compile and link a CUDA kernel...

**Comment by bentsherman** [comment]
...and it works! The example program now has both OpenCL and CUDA capabilities. I think this feature is ready to merge whenever we can get around to it.

**Comment by 4ctrl-alt-del** [comment]
Thank you very much for working on this and adding it to ACE. At this point ACE comes with an impressive list of supported features for accelerated computing...

OpenCL
CUDA
MPI

:)

**Comment by bentsherman** [comment]
Indeed it does! I'll wait for development to slow down on your end (probably after metadata and MPI) before merging.

**Comment by bentsherman** [comment]
Added two main features to this branch over the week:
- `ECUDAStream`: wrapper for CUDA streams that are necessary for doing things like overlapping kernel calls / memory transfers
- `ECUDAKernel`: parallel of `EOpenCLKernel` for CUDA. It turns out that there's a `cudaLaunchKernel()` function which enables OpenCL-style kernel execution. I couldn't get the function to work, but I figured out a hack that still makes this paradigm possible. If we can figure out how to use `cudaLaunchKernel()` without segfaulting then we can get rid of the hack.

**Comment by bentsherman** [comment]
Figured out how to use `cudaLaunchKernel()`, so the programming model for CUDA is a bit cleaner. I'm going to look into NVRTC, which provides runtime compilation and may allow us to accept CUDA kernel source code exactly as we do OpenCL source.

**Comment by bentsherman** [comment]
Duly noted. I figured out how to use NVRTC, so the CUDA interface matches the OpenCL interface very closely. I see where you're going with these new analytic classes and I like it! When everything is ready on your end I can backtrack and add the CUDA interface in one big commit; trying to merge at this point would be a nightmare.

**Comment by bentsherman** [comment]
I just pushed a new `cuda-support` branch which is now up to date with current ACE. @4ctrl-alt-del please take a look and test the example analytic when you can and let me know if there are any issues. As far as my end is concerned it is ready to merge.

**Comment by 4ctrl-alt-del** [comment]
I am rather overwhelmed with debugging right now and will be going on vacation at the end of this week. So this will have to be put on hold.
